File: tik/Lab5/Lab_5_TIC/HammingCode/ConvertText.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace HammingCode
{
    class ConvertText
    {
        public static Dictionary<char, double> table_2_Prob = new Dictionary<char, double>(2);
        public static Dictionary<char, double> table_3_Prob = new Dictionary<char, double>(4);
        public static Dictionary<char, double> table_4_Prob = new Dictionary<char, double>(8);
        public static Dictionary<char, double> table_5_Prob = new Dictionary<char, double>(16);

        public static Dictionary<char, BitArray> table_2_Code = new Dictionary<char, BitArray>(2);
        public static Dictionary<char, BitArray> table_3_Code = new Dictionary<char, BitArray>(4);
        public static Dictionary<char, BitArray> table_4_Code = new Dictionary<char, BitArray>(8);
        public static Dictionary<char, BitArray> table_5_Code = new Dictionary<char, BitArray>(16);
        public Dictionary<char, double> GetProbabiltys(string text)
        {
            if (text.Length == 0) return null;
            Dictionary<char, double> charsProbab = new Dictionary<char, double>();
            foreach(char c in text)
            {
                if(!charsProbab.Keys.Contains(c))
                {
                    charsProbab[c] = text.Count(el => el == c) / text.Length;
                }
            }
            return charsProbab;
        }

        public static Dictionary<char, double> GetDictionaryForTable3(int numberOfForm)
        {
            int COUNT_VARS = 0;
            object numberTable = null;
            int middleTable = 0;
            switch (numberOfForm)
            {
                case 3:
                    {
                        numberTable = table_2_Prob;
                        COUNT_VARS = 4;
                        middleTable = 2;
                    }
                    break;
                case 4:
                    {
                        numberTable = table_3_Prob;
                        COUNT_VARS = 8;
                        middleTable = 4;
                    } break;
                case 5:
                    {
                        numberTable = table_4_Prob;
                        COUNT_VARS = 16;
                        middleTable = 8;
                    } break;

            }

            Dictionary<char, double> newDict = new Dictionary<char, double>(COUNT_VARS);
            newDict.Clear();
            for (int i = 0; i < COUNT_VARS; i++)
            {
                if (i < COUNT_VARS / 2)
                {
                    newDict[char.Parse(Convert.ToString(i, 16))] = ((Dictionary<char, double>)numberTable)[char.Parse((i % middleTable).ToString())] * table_2_Prob['0'];
                }
                else
                {
                    newDict[char.Parse(Convert.ToString(i, 16))] = ((Dictionary<char, double>)numberTable)[char.Parse((i % middleTable).ToString())] * table_2_Prob['1'];
                }
            }
            
            switch (numberOfForm)
            {
                case 3: table_3_Prob = newDict; break;
                case 4: table_4_Prob = newDict; break;
                case 5: table_5_Prob = newDict; break;
            }
            return newDict;
        }

        public static Dictionary<char, double> NormalizeIndex(Dictionary<char, double> oldDict)
        {
            Dictionary<char, double> newDict = new Dictionary<char, double>();
            newDict.Clear();
            for(int i=0; i< oldDict.Count; i++)
            {
                newDict.Add(char.Parse(i.ToString()), oldDict.Values.ToArray()[i]);
            }
            string tm = "NormalizeIndex";
            for (int i =0;  i < oldDict.Count; i++)
            {
                tm += "Key:" + oldDict.Keys.ToArray()[i].ToString() + "|Value:" + oldDict.Values.ToArray()[i].ToString()+"\n";
            }
            MessageBox.Show(tm);
            return newDict;
        }
   }
}

File: lab_5/task_1/convertText.py
class ConvertText:
    def __init__(self):
        self.table_2_Prob = {}
        self.table_3_Prob = {}
        self.table_4_Prob = {}
        self.table_5_Prob = {}

        self.table_2_Code = {}
        self.table_3_Code = {}
        self.table_4_Code = {}
        self.table_5_Code = {}
        
    
    def GetProbabilities(self, text: str = "") -> dict:
        if(len(text) == 0):
            return None

        charsProb = {}
        for char in text:
            if(char not in charsProb.keys()):
                charsProb[char] = text.count(char) / len(text)
        
        return charsProb
    
    
    def GetDictionaryForTable3(self, numberOfForm: int = 0) -> dict:
        COUNTS_VAR = 0
        numberTable = None
        middleTable = 0
        
        match(numberOfForm):
            case 3:
                numberTable = self.table_2_Prob
                COUNTS_VAR = 4
                middleTable = 2
            case 4:
                numberTable = self.table_3_Prob
                COUNTS_VAR = 8
                middleTable = 4
            case 5:
                numberTable = self.table_3_Prob
                COUNTS_VAR = 8
                middleTable = 4
        
        newDict = {}
        for i in range(0,COUNTS_VAR):
            if(i < COUNTS_VAR / 2):
                newDict[str(i)] = numberTable[str(i%middleTable)] * self.table_2_Prob['0']
            else:
                newDict[str(i)] = numberTable[str(i%middleTable)] * self.table_2_Prob['1']
        
        match(numberOfForm):
            case 3:
                self.table_3_Prob = newDict
            case 4:
                self.table_4_Prob = newDict
            case 5:
                self.table_5_Prob = newDict
        
        return newDict


    def NormalizeIndex(self, oldDict: dict = {}) -> dict:
        newDict = {}
        for i in range(0, len(oldDict.keys())):
            newDict[str(i)] = list(oldDict.values())[i]
        tmp = "NormalizeIndex"

        a = list(oldDict.keys())
        b = list(oldDict.values())

        for i in range(0, len(oldDict.keys())):
            tmp += "Key: {0} | Value {1}\n".format(a[i], b[i])
        
        print(tmp)
        return newDict


result = ConvertText()




